Expose raw saved tensors for codegen functions

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60975
* #60663
* **#60565**

Summary: This PR is part of a stack aimed to address #58512.
Same as #60551 but for codegen functions.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466225](https://our.internmc.facebook.com/intern/diff/D29466225)

Varal7 committed:
Expose raw saved tensors for codegen functions

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

soulitzer commented on test/test_autograd.py:
```diff
@@ -4669,22 +4669,39 @@ def test_grad_fn_attr_bindings(self):
         self.assertIsInstance(out.grad_fn._saved_dim, int)
 
+        saved = out.grad_fn._raw_saved_tensors[0]
+        saved.register_hooks(lambda x: x, lambda x: x)
```
Isn't this undefined behavior if Python still references `saved` after its C++ freed? Maybe we should delete after we register hooks here

Varal7 replied:
Yes you're right

soulitzer commented on test/test_autograd.py:
```diff
@@ -4669,22 +4669,39 @@ def test_grad_fn_attr_bindings(self):
+            saved[0].register_hooks(lambda x: x, lambda x: x)
+        out.sum().backward()
+        with self.assertRaisesRegex(RuntimeError, "after it has been freed"):
+            saved[1].register_hooks(lambda x: x, lambda x: x)
```
same for here

soulitzer replied:
we should probably just not do this test, and instead add a comment above the `del saved` on the lifetime issue

Varal7 replied:
My plan is to remove this test for cases 2 and 3 of the quip I sent you privately (so whenever we deal with a  `std::vector<SavedVariable>`) but keep it in case 1 because we're not actually deleting the `SavedVariable` during backward (only calling `.reset_data()` on it).

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466225).

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466225](https://our.internmc.facebook.com/intern/diff/D29466225)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466225).

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466225](https://our.internmc.facebook.com/intern/diff/D29466225)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466225).

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466225](https://our.internmc.facebook.com/intern/diff/D29466225)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466225).

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466225](https://our.internmc.facebook.com/intern/diff/D29466225)

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for codegen functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466225](https://our.internmc.facebook.com/intern/diff/D29466225)

[ghstack-poisoned]
